### PR TITLE
Publicize: google plus preview

### PR DESCRIPTION
**Pull request by jhnstn**
This adds google plus to the publicize preview

screen shot 2017-05-17 at 12 36 06 pm

**Testing**
- Enable publicize sharing on `calypso.localhost:3000` : `ENABLE_FEATURES=publicize-scheduling make run`
- Authorize google plus on a premium site
- Open the sharing tab on a published post
- Click 'Preview'
- Google+  should be available for previewing
- The Google+ should display the post title, excerpt, and featured image ( if present ) and the google user account info

_breaks up #12644_

### CODE REVIEW

**Comment by jhnstn** [comment]
Copied over from @iamtakashi comment:

---
**Google Plus**
* We need to show the summery and the title of the post.

image

File: client/components/vertical-menu/items/social-item.jsx [hunk 1]
```diff
@@ -13,6 +13,7 @@ import SocialLogo from 'social-logos';
 const services = translate => ( {
 	facebook: { icon: 'facebook', label: translate( 'Facebook feed' ) },
 	google: { icon: 'google', label: translate( 'Google search' ) },
+	google_plus: { icon: 'google-plus', label: translate( 'Google+ ' ) },
```
Comment (a8ci18n) [inline client/components/vertical-menu/items/social-item.jsx]: Hi! I've found a possible matching string that has already been translated **55** times:
`translate( 'Google' )` *ES Score: 11*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/publicize-google-plus-preview)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/components/share/google-plus-share-preview/style.scss [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+.google-plus-share-preview__message,
+.google-plus-share-preview__article-summary {
+	font-size: 14px;
+	line-height: 20px;
+	margin: 0 16px 16px;
+	word-break: break-word;
+}
+
+.google-plus-share-preview__article-title {
+	border-top: 1px solid #ebebeb;
```
Comment (retrofox) [inline client/components/share/google-plus-share-preview/style.scss]: do we have a sass constant for this?

File: client/blocks/sharing-preview-pane/index.jsx [hunk 3]
```diff
@@ -34,7 +35,10 @@ class SharingPreviewPane extends PureComponent {
 	};
 
 	static defaultProps = {
-		services: [ 'facebook', 'twitter' ]
+		services: [
+			'facebook',
+			'google_plus',
+			'twitter' ]
```
Comment (retrofox) [inline client/blocks/sharing-preview-pane/index.jsx]: maybe close the bracket on its own line?

File: client/components/share/google-plus-share-preview/style.scss (hunk 2 above)
Comment (jhnstn) [inline client/components/share/google-plus-share-preview/style.scss]: I couldn't find any. @iamtakashi do you think we should add one?

**Comment by sirbrillig** [comment]
Tested according to instructions. I get a blank preview in Google+ and an error in the console:

screen shot 2017-05-17 at 6 11 32 pm

screen shot 2017-05-17 at 6 12 55 pm

**Comment by jhnstn** [comment]
@sirbrillig  ahh I think I know what that is. Was your post empty?

**Comment by sirbrillig** [comment]
It's a new site (premium upgrade) with just a single Headstart-generated post, viewing in an Incognito window (no previously saved state).

More digging reveals that the `post` prop is null, even though the `postId` is set correctly. That is set by the `getSitePost()` selector, which looks for a post in `state.posts`. Here's what `state.posts` looks like:

screen shot 2017-05-17 at 6 52 20 pm

It's pretty much empty, which is why we have no post in the component. (There is some `count.requesting` stuff that I'm not entirely clear on.)

Also worth noting that none of the other previews work either, even though they don't throw errors.

screen shot 2017-05-17 at 6 57 34 pm

File: client/components/share/google-plus-share-preview/style.scss (hunk 2 above)
Comment (iamtakashi) [inline client/components/share/google-plus-share-preview/style.scss]: I don't think it's worth it. The styles for this previews won't be shared with anywhere else.

**Comment by iamtakashi** [comment]
I wasn't sure if this was ready for review but I'm noting here just in case.

> The Google+ should display the post title, excerpt, and featured image ( if present ) and the google user account info

It looks like the message you type when you share the post is missing. The message should appear above the excerpt with a line break. ("Do you have a trip coming up?" is the example in my mockup)

File: client/components/share/google-plus-share-preview/style.scss (hunk 2 above)
Comment (retrofox) [inline client/components/share/google-plus-share-preview/style.scss]: mmm, maybe this preview stuff won't be shared but what could be shared is the _meaning_ of the color. There is a rule for this: 
> Don't use custom colors, always utilize what _colors.scss provides. If you have to set a color, use lowercase hex values, and shorten them to their smallest expression (like #aaa).

reference: https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md#the-dry-principle

Comment (iamtakashi) [inline client/components/share/google-plus-share-preview/style.scss]: I'm aware of the piece of the guideline. But, in this case, we're just replicating their UI styles here, which means it's not really something to do with Calypso UI, so I'm not sure if it has any meaning to the value from Calypso design point of view.

File: client/components/share/helpers.js [hunk 4]
```diff
@@ -20,3 +20,20 @@ export const hardTruncation = limit => title => title.slice( 0, limit ).concat(
 
 export const firstValid = ( ...predicates ) => a =>
 	find( predicates, ( p => false !== p( a ) ) )( a );
+
+export const truncateArticleContent = ( maxCharacters, content ) => {
```
Comment (jhnstn) [inline client/components/share/helpers.js]: I know this is very close to `truncateAtSpace` above but the that function has the odd effect of making you _guess_ the length of the last word with the `lower` param. if `lower` and `upper` are too close, you don't get any truncation. If they are too far apart you get an unnecessary ellipsis.

This was copied over from SEO so I will prune/refactor after we get the share feature in a good stable place

**Comment by sirbrillig** [comment]
I now get this error `Uncaught TypeError: Cannot read property 'length' of null at truncateArticleContent`, but in Slack we discussed that maybe the pre-generated content is the cause of this issue and you may deal with that in a future PR. Since this is behind a feature flag, I'd say it's probably safe to merge. If you understand the underlying problem, maybe just create an Issue for it so there's something to link to?

**Pull request merged**